segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/extendeds3/ExtendedS3Storage.java

```diff
@@ -0,0 +1,445 @@
+        }
+
+        S3ObjectMetadata metadata = new S3ObjectMetadata();
+        metadata.setContentLength((long) 0);
+
+        PutObjectRequest request = new PutObjectRequest(config.getBucket(),
+                config.getRoot() + streamSegmentName,
+                (Object) null);
+
+        AccessControlList acl = new AccessControlList();
+        acl.addGrants(new Grant[]{
+                new Grant(new CanonicalUser(config.getAccessKey(), config.getAccessKey()),
+                        Permission.FULL_CONTROL)
+        });
+
+        request.setAcl(acl);
+
+        client.putObject(request);
+
+        LoggerHelpers.traceLeave(log, "create", traceId);
+        return syncGetStreamSegmentInfo(streamSegmentName);
+    }
+
+    private Void syncWrite(SegmentHandle handle, long offset, InputStream data, int length) throws StreamSegmentSealedException, BadOffsetException {
+        Preconditions.checkArgument(!handle.isReadOnly(), "handle must not be read-only.");
+
+        long traceId = LoggerHelpers.traceEnter(log, "write", handle.getSegmentName(), offset, length);
+
+        SegmentProperties si = syncGetStreamSegmentInfo(handle.getSegmentName());
+
+        if (si.isSealed()) {
+            throw new StreamSegmentSealedException(handle.getSegmentName());
+        }
+
+        if (si.getLength() != offset) {
+            throw new BadOffsetException(handle.getSegmentName(), si.getLength(), offset);
+        }
+
+        client.putObject(this.config.getBucket(), this.config.getRoot() + handle.getSegmentName(),
+                Range.fromOffsetLength(offset, length), data);
+        LoggerHelpers.traceLeave(log, "write", traceId);
+        return null;
+    }
+
+    private Void syncSeal(SegmentHandle handle) {
+        Preconditions.checkArgument(!handle.isReadOnly(), "handle must not be read-only.");
+
+        long traceId = LoggerHelpers.traceEnter(log, "seal", handle.getSegmentName());
+
+        AccessControlList acl = client.getObjectAcl(config.getBucket(),
+                config.getRoot() + handle.getSegmentName());
+        acl.getGrants().clear();
+        acl.addGrants(new Grant[]{new Grant(new CanonicalUser(config.getAccessKey(), config.getAccessKey()),
+                Permission.READ)});
+
+        client.setObjectAcl(
+                new SetObjectAclRequest(config.getBucket(), config.getRoot() + handle.getSegmentName()).withAcl(acl));
+        LoggerHelpers.traceLeave(log, "seal", traceId);
+        return null;
+    }
+
+    /**
+     * The concat is implemented using extended S3 implementation of multipart copy API. Please see here for more details on multipart copy.
+     * http://docs.aws.amazon.com/AmazonS3/latest/dev/CopyingObjctsUsingLLJavaMPUapi.html
+     * The multipart copy is an atomic operation. We schedule two parts and commit the atomically using completeMultiPartUpload call.
+     * The first part is the destination object and the second part is the source segment.
```

Comment:
You use target below instead of destination, let's use terminology consistently.